### ISSUE TITLE
IoU threshold values clarification

### ISSUE THREAD

**Issue by eldarkurtic** [ultralytics/yolov5#3907]
Hi @glenn-jocher , I have a few questions about some IoU threshold values used in the code.

1. Evaluation epochs (`test.run(...)`) are using the default value of `iou_thres=0.6,  # NMS IoU threshold`, but the metric is labelled as `mAP@0.5`. Is this a typo or? I think the original `darknet` version and the paper are using `iou_thres=0.5`. Also, smaller `iou_thres` should increase the accuracy a bit, since it's "easier to predict correctly".
2. Confusion matrix used in the `test.run(...)` is initialized with the default value of `iou_thres=0.45`. Why is this not the same as the `iou_thres` that is given to the `test.run(...)` function?
3. After the full training phase, the evaluation of `best` and `last` checkpoints is triggered with the call to the `test.run(...)` with `iou_thres=0.7` value. Why is this not the same as in the evaluation during the training, since the reported accuracy of the same model evaluated during the training (with `iou_thres=0.6`) and after the training (with `iou_thres=0.7`) won't be the same.

Update:
4. Could you also please clarify these IoU values in the table description, the columns says `mAP@0.5` but the detailed explanation is using some completely different values?
<img width="870" alt="Screenshot 2021-07-06 at 19 23 17" src="https://user-images.githubusercontent.com/8884008/124642747-7d133600-de90-11eb-8493-7441f2e46a83.png">

**Comment by daniel-toth-leeder** [comment]
￼ 😒 my    
🐂__
**👩‍👦**🐽

**Comment by eldarkurtic** [comment]
Hi @glenn-jocher , I have been going through the code a bit and realized that the `iou` and `iou_thresh` are actually IoUs used for the non-max suppression and not for the mAP calculation. Given that, I would update these questions in the following way (please ignore the previous ones):

1. Looks like the `darknet`source code for validation on COCO is using `0.5` for the NMS . (according to https://github.com/pjreddie/darknet/blob/b339a70cbb8f4378b2bc19e560e9db8a6cde3737/examples/coco.c#L155 and https://github.com/pjreddie/darknet/blob/b339a70cbb8f4378b2bc19e560e9db8a6cde3737/examples/coco.c#L198) 
Is there a specific reason why is this changed in the YoloV5 to `0.6`/`0.7`? Do these new values "work better" or ?

2. Confusion matrix used in the `test.run(...)` is initialized with the default value of `iou_thres=0.45`. Why this isn't `0.5`, so that the generated plots are consistent with the metric we are looking at (`mAP@0.5`)?

3. After the full training phase, the evaluation of best and last checkpoints is triggered with the call to the `test.run(...)` with `iou_thres=0.7` value for the NMS. Why is this not the same as in the evaluation during the training, since the reported accuracy of the same model evaluated during the training (with `iou_thres=0.6`) and after the training (with `iou_thres=0.7`) most likely won't be the same?

4. This question is completely irrelevant given that `--iou` is used for the NMS, and not for metric calculation.

**Comment by glenn-jocher** [comment]
@eldarkurtic 

1. higher mAP
2. iou_thres=0.45 is inference default for detect.py, hub and confusion matrix
3. 0.6 tends to produce better results at mAP@0.5, 0.7 tends to produce better results on mAP@0.5:0.95 metric

**Comment by eldarkurtic** [comment]
3. But shouldn't these values be exactly the same? We train a model, during the training we evaluate it with `iou_thresh=0.6` and based on results obtained with this value we pick the best model. Later on, after the last training epoch, we run the final evaluation to check the best model we picked (based on the `0.6` value) and we get completely different results from the ones we picked as "the best" during the training.

**Comment by glenn-jocher** [comment]
@eldarkurtic 0.6 is the default in train and test, there is no 'different results'.

**Comment by eldarkurtic** [comment]
During the training `test.run` is called with the default value `iou_thresh=0.6`: https://github.com/ultralytics/yolov5/blob/8c6f9e15bfc0000d18b976a95b9d7c17d407ec91/train.py#L389-L400
https://github.com/ultralytics/yolov5/blob/8c6f9e15bfc0000d18b976a95b9d7c17d407ec91/test.py#L37

And then, after the training phase, there is an evaluation of the `best` and `last` models, which is invoked with `iou_thresh=0.7`:
https://github.com/ultralytics/yolov5/blob/8c6f9e15bfc0000d18b976a95b9d7c17d407ec91/train.py#L457-L467

**Comment by glenn-jocher** [comment]
@eldarkurtic ah this is only applies to the narrow use case of full COCO dataset training, so no other datasets will see this.

**Comment by eldarkurtic** [comment]
Yes, I have noticed it since I am running Yolo on the full COCO dataset :) . Should I submit a PR to change this back to the default `0.6` value?

**Comment by glenn-jocher** [comment]
@eldarkurtic ah sure, please submit a PR!

**Comment by glenn-jocher** [comment]
@eldarkurtic good news 😃! Your original issue may now be fixed ✅ in PR #3934. To receive this update:
- **[Git](https://github.com/ultralytics/yolov5)** – `git pull` from within your `yolov5/` directory or `git clone https://github.com/ultralytics/yolov5` again
- **[PyTorch Hub](https://pytorch.org/hub/ultralytics_yolov5/)** – Force-reload with `model = torch.hub.load('ultralytics/yolov5', 'yolov5s', force_reload=True)`
- **[Notebooks](https://github.com/ultralytics/yolov5/blob/master/tutorial.ipynb)** – View updated notebooks  <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **[Docker](https://hub.docker.com/r/ultralytics/yolov5)** – `sudo docker pull ultralytics/yolov5:latest` to update your image <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>

Thank you for spotting this issue and informing us of the problem. Please let us know if this update resolves the issue for you, and feel free to inform us of any other **issues** you discover or **feature requests** that come to mind. Happy trainings with YOLOv5 🚀!

**Comment by wtjasmine** [comment]
Hi @glenn-jocher, I have a question regarding the iou_thres in the YOLOv5 model During training, the YOLOv5 model is optimized with an iou_thres of NMS of 0.6, and the best model is chosen based on this threshold during the NMS process. However, in the detect.py script, the default iou_thres is set to 0.45 during inference. I noticed this difference and was wondering if choosing another value for iou_thres during validation or detection would introduce inconsistency in the model's performance. Could you provide insights on the rationale behind this difference and whether adjusting iou_thres during inference may affect the model's consistency with the training process? Thank you for your time and assistance.

**Comment by glenn-jocher** [comment]
Hi there! Great question! The `iou_thres` used during training (0.6) and the default in `detect.py` (0.45) are indeed different. This is primarily to balance precision and recall during different stages. A higher `iou_thres` during training helps in selecting the best model by being stricter in the overlap criterion, which generally aids in better model discrimination.

During inference, a lower `iou_thres` can help in maintaining a higher recall, ensuring fewer true positives are missed, which might be more desirable in practical applications.

Adjusting the `iou_thres` during inference can indeed affect performance consistency compared to training, as it may alter the balance between precision and recall. You can experiment with this threshold to find the best setting for your specific use case. Happy experimenting! 😊
